source/exe/win32/main.cc

```diff
@@ -0,0 +1,74 @@
+#include <string>
+
+#include "common/common/thread_impl.h"
+
+#include "exe/main_common_base.h"
+
+#include "absl/debugging/symbolize.h"
+
+// clang-format off
+#include <winsock2.h>
+#include <ws2tcpip.h>
+// clang-format on
+
+// NOLINT(namespace-envoy)
+
+/**
+ * Basic Site-Specific main()
+ *
+ * This should be used to do setup tasks specific to a particular site's
+ * deployment such as initializing signal handling. It calls main_common
+ * after setting up command line options.
+ */
+
+namespace {
+class WinsockCleanup {
+public:
+  ~WinsockCleanup() { WSACleanup(); }
+};
+} // namespace
+
+int main(int argc, char** argv) {
+  // absl::Symbolize mostly works without this, but this improves corner case
+  // handling, such as running in a chroot jail.
+  absl::InitializeSymbolizer(argv[0]);
+
+  const WORD wVersionRequested = MAKEWORD(2, 2);
+  WSADATA wsaData;
+
+  const int rc = WSAStartup(wVersionRequested, &wsaData);
+  if (rc != 0) {
+    printf("WSAStartup failed with error: %d\n", rc);
+    return 1;
+  }
+  WinsockCleanup winsock_cleanup;
+
+  std::unique_ptr<Envoy::MainCommonBase> main_common_base;
+  std::unique_ptr<Envoy::OptionsImpl> options;
+  Envoy::Event::RealTimeSystem real_time_system;
+  Envoy::DefaultTestHooks default_test_hooks;
+  Envoy::ProdComponentFactory prod_component_factory;
+  Envoy::Thread::ThreadFactoryImplWin32 thread_factory;
+  auto hotRestartVersion = [](uint64_t, uint64_t, bool) -> std::string { return "disabled"; };
```

Comment:
I think I like the way this is going, but there's pretty significant swaths of common code that is replicated. Would you consider factoring that out and putting it in main_common_base.cc?

Reply:
I still think we can factor out more, but I also understand the need to move this PR forward. Can we have a TODO here to factor out the common parts of this with the posix version? 

Maybe the right thing is to move all that into MainCommonBase as an alternate constructor. In any case, please don't iterate more on that now, let's follow up. There's tons of great stuff in here.

But I think you are going to have to do this refactoring rather than someone else, as no one else will be able to touch this code unless CI for Windows is added.

Reply:
I think we should clean this up a bit more before shipping. Specifically, this PR introduces some duplication and a drift in behavior for common parts of main. See below for the exception handling, which is more fine grained here, but probably also should be in the POSIX version as it's not tied to platform.

I think we should just have some hooks for platform initialization that get called, and most of this file can be the same.

Reply:
I think the exception handling is the same in both versions -- see https://github.com/greenhouse-org/envoy/blob/37c183286611023e2fac690cb5f1d149d02a3aae/source/exe/posix/main.cc#L25-L33 for the POSIX version.

The exception handling is different if the `main_common` function: https://github.com/greenhouse-org/envoy/blob/37c183286611023e2fac690cb5f1d149d02a3aae/source/exe/posix/main_common.cc#L53 is used, though